pkg/api/message/v1/service.go

```diff
@@ -35,6 +36,24 @@ const (
 
 	// 1048576 - 300 - 62 = 1048214
 	MaxMessageSize = pubsub.DefaultMaxMessageSize - MaxContentTopicNameSize - 62
+
+	// maxRequestLimitPerBatch defines the maximum number of request we can support per batch.
+	maxRequestLimitPerBatch = 50
+
+	// maxTopicsPerRequest defines the maximum number of topics that can be queried in a single request.
+	maxTopicsPerRequest = 1024
+
+	// maxTopicsPerBatch defines the maximum number of topics that can be queried in a batch query. This
+	// limit is imposed in additional to the per-query limit maxTopicsPerRequest.
+	maxTopicsPerBatch = 4096
+
+	// maxConcurrentSubscribedTopics is the total number of subscribed topics currently being registered by our node.
+	// request to surpass that limit would be rejected, as the server would not be able to scale to that limit.
+	maxConcurrentSubscribedTopics = 1024 * 1024
```

Comment:
This depends on the available memory in the node, which changes depending on the deployment environment (devnet vs prod). Currently in prod we've seen over 6M concurrent topic subscriptions before memory alerts start firing.

Alternatively, or in additional to this, I think it would be great to limit concurrent topic subscriptions by IP. I think I'd prefer to start with that vs a global limit across all IPs, but I'm not against having a global limit either, it just might need to be configurable via CLI option instead of a const.

Curious about @neekolas's thoughts on this too.

Reply:
Yes, I agree - we need multiple-tiers of limits to ensure the service won't be disrupted due to 1% of the clients placing unrealistic load on the server. I'm very much open to any other proposed default, which we would be able to dial-in later on. Does 100M sounds better as a starting point ?

Reply:
i.e. use a const for default, and override it via .env/config file/cli.

Reply:
I was thinking of much more fundamental change to address this level of scale. having go-routine per subscription is ridiculous and won't scale. Instead, I think that we should do the following:
1. create a single subscription per node, which would receive *all* the messages.
2. for each subscription (regardless of it's size), create a matching topics map.
3. when a message received, compare it to all the maps, after Unmarshal'ing it *once*, and dispatch it to all the recipients.

this approach is less efficient in quiescent scenarios, but would create a limit on the rendered dynamic system load.

Reply:
It's going to be very difficult to determine what the right global limit is. Rejecting subscription requests might avoid one incident (nodes restarting), but it will create another incident (subscriptions mysteriously failing). 

Per IP, we can at least get a sense of what is reasonable and put a cap in so that 99th percentile user doesn't break the network for everyone.

Reply:
the updated code doesn't reject any subscription requests anymore.